velox/exec/tests/TaskTest.cpp

```diff
@@ -615,6 +617,191 @@ TEST_F(TaskTest, singleThreadedCrossJoin) {
+      continueFuture_ = externalBlocker_->continueFuture();
+      return nullptr;
+    }
+    auto temp = std::move(input_);
+    input_ = nullptr;
+    return temp;
+  }
+
+  exec::BlockingReason isBlocked(ContinueFuture* future) override {
+    if (continueFuture_.valid()) {
+      *future = std::move(continueFuture_);
+      return exec::BlockingReason::kWaitForConsumer;
+    }
+    return exec::BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return noMoreInput_;
+  }
+
+ private:
+  RowVectorPtr input_;
+  ExternalBlocker* externalBlocker_;
+  folly::SemiFuture<folly::Unit> continueFuture_;
+};
+
+class TestExternalBlockableTranslator
+    : public exec::Operator::PlanNodeTranslator {
+  std::unique_ptr<exec::Operator> toOperator(
+      exec::DriverCtx* ctx,
+      int32_t id,
+      const core::PlanNodePtr& node) override {
+    if (auto tempNode =
+            std::dynamic_pointer_cast<const TestExternalBlockableNode>(node)) {
+      return std::make_unique<TestExternalBlockableOperator>(id, ctx, tempNode);
+    }
+    return nullptr;
+  }
+};
+
+TEST_F(TaskTest, singleThreadedExecutionExternalBlockable) {
+  exec::Operator::registerOperator(
+      std::make_unique<TestExternalBlockableTranslator>());
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>(1'000, [](auto row) { return row; }),
+  });
+  auto blocker = std::make_shared<ExternalBlocker>();
+  // Filter + Project.
+  auto plan =
+      PlanBuilder()
+          .values({data, data, data})
+          .addNode([blocker](std::string id, core::PlanNodePtr input) mutable {
+            return std::make_shared<TestExternalBlockableNode>(
+                id, input, std::move(blocker));
+          })
+          .project({"c0"})
+          .planFragment();
+
+  ContinueFuture continueFuture = ContinueFuture::makeEmpty();
+  // First pass, we don't activate the external blocker, expect the task to run
+  // without being blocked.
+  auto task = std::make_shared<exec::Task>(
+      "single.execution.task.0", plan, 0, std::make_shared<core::QueryCtx>());
+  std::vector<RowVectorPtr> results;
+  for (;;) {
+    auto result = task->next(&continueFuture);
+    if (!result) {
+      break;
+    }
+    EXPECT_FALSE(continueFuture.valid());
+    results.push_back(std::move(result));
+  }
+  EXPECT_EQ(3, results.size());
+  // Second pass, we will now use external blockers to block the task.
+  auto task2 = std::make_shared<exec::Task>(
+      "single.execution.task.1", plan, 0, std::make_shared<core::QueryCtx>());
+  results.clear();
+  continueFuture = ContinueFuture::makeEmpty();
+  // Before we block, we expect `next` to get data normally.
+  results.push_back(task2->next(&continueFuture));
+  EXPECT_TRUE(results.back() != nullptr);
+  // Now, we want to block the pipeline by external event. We expect `next` to
+  // return null.  The `future` should be updated for the caller to wait before
+  // calling next() again
+  blocker->isBlocked = true;
```

Comment:
Is it safe to access block here as it has already been moved to the first nonblocking test at l756? Consider to change it to a unqiue ptr and let the test thread to own the blocker object or just copy the shared pointer when pass it to the plan node.